ksqldb-engine/src/main/java/io/confluent/ksql/schema/ksql/inference/DefaultSchemaInjector.java

```diff
@@ -185,10 +177,24 @@ private SchemaAndId getSchema(
     return result.schemaAndId.get();
   }
 
-  private static void validateSchemaInference(
+  private static boolean shouldInferSchema(
+      final Optional<Integer> schemaId,
+      final ConfiguredStatement<CreateSource> statement,
       final FormatInfo formatInfo,
       final boolean isKey
   ) {
+    final boolean hasTableElements =
+        isKey ? hasKeyElements(statement) : hasValueElements(statement);
+
+    /*
+     * Conditions for schema inference:
+     *   1. key_schema_id or value_schema_id property exist or
+     *   2. Table elements doesn't exist and format support schema inference
+     */
+    if (!schemaId.isPresent() && (hasTableElements || !formatSupportsSchemaInference(formatInfo))) {
```

Comment:
Can we break this apart now that it's in its own method (thanks for doing that!) so it's more readable? For example:
```
if (schemaId.isPresent()) {
    // do validation specific to this case
    return true;
} else {
    // ... etc ...
}
```